internal/ledgerutil/verify/verify.go

```diff
@@ -0,0 +1,228 @@
+/*
+Copyright Hitachi, Ltd. 2023 All rights reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package verify
+
+import (
+	"bytes"
+	"fmt"
+	"path/filepath"
+
+	"github.com/hyperledger/fabric-protos-go/common"
+	"github.com/hyperledger/fabric/common/ledger/blkstorage"
+	"github.com/hyperledger/fabric/common/metrics/disabled"
+	"github.com/hyperledger/fabric/core/ledger/kvledger"
+	"github.com/hyperledger/fabric/internal/fileutil"
+	"github.com/hyperledger/fabric/internal/ledgerutil/jsonrw"
+	"github.com/hyperledger/fabric/protoutil"
+	"github.com/pkg/errors"
+)
+
+const (
+	ledgersDataDirName      = "ledgersData"
+	blockResultJsonFilename = "blocks.json"
+
+	errorDataHashMismatch = "DataHash mismatch"
+	errorPrevHashMismatch = "PreviousHash mismatch"
+)
+
+// VerifyLedger - Verifies the integrity of ledgers in a single peer
+// This function verifies the ledgers specified in blockStorePath, specifically the integrity of
+// the hash value (DataHash) and the previous hash value (PreviousHash) in every block.
+// It creates a directory in outputDir for each ledger and stores the check result as a JSON file
+// in the directory.
+func VerifyLedger(blockStorePath string, outputDir string) (bool, error) {
+	// Save whether any error is detected
+	ledgerErrorFound := false
+
+	// Get the block store provider
+	blockStoreProvider, err := getBlockStoreProvider(blockStorePath)
+	if err != nil {
+		return false, err
+	}
+	defer blockStoreProvider.Close()
+
+	// Retrieve the list of the ledgers (i.e. channels) available
+	list, err := blockStoreProvider.List()
+	if err != nil {
+		return false, err
+	}
+
+	// Iterate over the channels
+	for _, ledgerId := range list {
+		// Create Output Directory
+		outputDirName := fmt.Sprintf("%s_verification_result", ledgerId)
+		outputDirPath := filepath.Join(outputDir, outputDirName)
+
+		empty, err := fileutil.CreateDirIfMissing(outputDirPath)
+		if err != nil {
+			return false, err
+		}
+		if !empty {
+			return false, errors.Errorf("%s already exists in %s. Choose a different location or remove the existing results. Aborting verify", outputDirName, outputDir)
+		}
+
+		// Open block provider
+		store, err := blockStoreProvider.Open(ledgerId)
+		if err != nil {
+			return false, err
+		}
+		defer store.Shutdown()
+
+		info, err := store.GetBlockchainInfo()
+		if err != nil {
+			return false, err
+		}
+
+		iterator, err := store.RetrieveBlocks(0)
```

Comment:
If a channel has been bootstapped by a snapshot, the starting block would be a non-zero. You can use the `info` struct in line 75 above to determine this, and get the first available block number.